File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solve_puzzle(level, diffs, times):
    solve_time = 0
    for i in range(len(diffs)):
        if diffs[i] <= level:
            solve_time += times[i]
        else:
            solve_time += (diffs[i] - level) * (times[i] + times[i-1]) + times[i]

    return solve_time


def solution(diffs, times, limit):
    level_arr = [i for i in range(min(diffs), max(diffs)+1)]

    # binary_search(level_arr, 0, len(level_arr)-1, diffs, times, limit)
    start, end = 0, len(level_arr)-1

    levelable = []
    while start <= end:
        pivot = (start + end) // 2

        if solve_puzzle(level_arr[pivot], diffs, times) <= limit:
            levelable.append(level_arr[pivot])
            end = pivot - 1
        else:
            start = pivot + 1

    answer = min(levelable)

    return answer

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    const isPossible = (level) => {
        let totalTime = 0;
        const n = diffs.length;
        
        for(let i = 0 ; i < n ; i ++){
            const diff = diffs[i];
            const timeCur = times[i];
            const timePrev = i === 0 ? 0 : times[i-1];
            
            if(diff <= level) totalTime += timeCur;
            else {
                const mistake = diff - level;
                const retryTime = (timeCur + timePrev) * mistake;
                totalTime += retryTime + timeCur;
            }
            
            if(totalTime > limit) return false;
        }
        return true;
    }
    
    let left = 1;
    let right = 100000;
    let result = 0;
    
    while(left <= right){
        const mid = Math.floor((left + right) / 2);
        if(isPossible(mid)){
            result = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    
    return result;
}